pkg/statistics/region_collection.go

```diff
@@ -222,44 +222,52 @@ func (r *RegionStatistics) Observe(region *core.RegionInfo, stores []*core.Store
 	// Better to make sure once any of these conditions changes, it will trigger the heartbeat `save_cache`.
 	// Otherwise, the state may be out-of-date for a long time, which needs another way to apply the change ASAP.
 	// For example, see `RegionStatsNeedUpdate` above to know how `OversizedRegion` and `UndersizedRegion` are updated.
-	conditions := map[RegionStatisticType]bool{
-		MissPeer:    len(peers) < desiredReplicas,
-		ExtraPeer:   len(peers) > desiredReplicas,
-		DownPeer:    len(downPeers) > 0,
-		PendingPeer: len(pendingPeers) > 0,
-		OfflinePeer: func() bool {
-			for _, store := range stores {
-				if store.IsRemoving() {
-					peer := region.GetStorePeer(store.GetID())
-					if peer != nil {
-						return true
-					}
-				}
+	var conditions RegionStatisticType
+	if len(peers) < desiredReplicas {
+		conditions |= MissPeer
+	}
+	if len(peers) > desiredReplicas {
+		conditions |= ExtraPeer
+	}
+	if len(downPeers) > 0 {
+		conditions |= DownPeer
+	}
+	if len(pendingPeers) > 0 {
+		conditions |= PendingPeer
+	}
+	for _, store := range stores {
+		if store.IsRemoving() {
+			peer := region.GetStorePeer(store.GetID())
+			if peer != nil {
+				conditions |= OfflinePeer
+				break
 			}
-			return false
-		}(),
-		LearnerPeer:      len(learners) > 0,
-		EmptyRegion:      regionSize <= core.EmptyRegionApproximateSize,
-		OversizedRegion:  region.IsOversized(regionMaxSize, regionMaxKeys),
-		UndersizedRegion: region.NeedMerge(maxMergeRegionSize, maxMergeRegionKeys),
-		WitnessLeader:    leaderIsWitness,
+		}
+	}
+	if len(learners) > 0 {
+		conditions |= LearnerPeer
+	}
+	if regionSize <= core.EmptyRegionApproximateSize {
+		conditions |= EmptyRegion
+	}
+	if region.IsOversized(regionMaxSize, regionMaxKeys) {
+		conditions |= OversizedRegion
+	}
+	if region.NeedMerge(maxMergeRegionSize, maxMergeRegionKeys) {
+		conditions |= UndersizedRegion
+	}
+	if leaderIsWitness {
+		conditions |= WitnessLeader
 	}
 	// Check if the region meets any of the conditions and update the corresponding info.
 	regionID := region.GetID()
-	for typ, c := range conditions {
-		if c {
-			info := r.stats[typ][regionID]
-			if typ == DownPeer {
-				if info == nil {
-					info = &RegionInfoWithTS{}
-				}
-				if info.(*RegionInfoWithTS).startDownPeerTS != 0 {
-					regionDownPeerDuration.Observe(float64(time.Now().Unix() - info.(*RegionInfoWithTS).startDownPeerTS))
-				} else {
-					info.(*RegionInfoWithTS).startDownPeerTS = time.Now().Unix()
-					logDownPeerWithNoDisconnectedStore(region, stores)
-				}
-			} else if typ == MissPeer {
+	for i := 0; i < len(regionStatisticTypes); i++ {
+		condition := RegionStatisticType(1 << i)
+		if conditions&condition != 0 {
```

Comment:
Prefer writing like this to prevent a big nest:

```suggestion
		if conditions&condition == 0 {
			continue
		}
		// Switch logic
```